Model/Product/Url/Builder.php

```diff
@@ -0,0 +1,110 @@
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @author Nosto Solutions Ltd <contact@nosto.com>
+ * @copyright 2017 Nosto Solutions Ltd
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD 3-Clause
+ *
+ */
+
+namespace Nosto\Tagging\Model\Product\Url;
+
+use Magento\Catalog\Model\Product;
+use Magento\CatalogUrlRewrite\Model\ProductUrlRewriteGenerator;
+use Magento\Framework\DataObject;
+use Magento\Framework\Filter\FilterManager;
+use Magento\Framework\Url;
+use Magento\UrlRewrite\Model\UrlFinderInterface;
+use Magento\UrlRewrite\Service\V1\Data\UrlRewrite;
+
+/**
+ * Url builder class cannibalised from the Magento core. When trying to get the URL of a product
+ * from the Magento backend, the DI rules inject inject a reference to \Magento\Backend\Model\Url
+ * instead of \Magento\Framework\Url. Both these classes implement the Magento\Framework\UrlInterface
+ * and are responsible for URL building. The building of the route parameters is the same whether
+ * it occurs on the backend or the frontend but the actual building of the URL differs from the
+ * backend to the frontend.
+ * <br />
+ * There's no clean way of changing this behaviour without modifying the core so this class contains
+ * code from \Magento\Catalog\Model\Product\Url which now always uses the frontend version of the
+ * Magento\Framework\UrlInterface class.
+ *
+ * @package Nosto\Tagging\Model\Product\Url
+ */
+class Builder extends DataObject
+{
+    private $data;
+    private $filter;
+    private $urlFinder;
+    private $urlFactory;
+
+    /**
+     * @param Url $urlFactory
+     * @param FilterManager $filter
+     * @param UrlFinderInterface $urlFinder
+     * @param array $data
+     */
+    public function __construct(
+        Url $urlFactory,
+        FilterManager $filter,
+        UrlFinderInterface $urlFinder,
+        array $data = []
+    ) {
+        parent::__construct($data);
+        $this->data = $data;
+        $this->filter = $filter;
+        $this->urlFinder = $urlFinder;
+        $this->urlFactory = $urlFactory;
+    }
+
+    public function getUrlInStore(Product $product, $storeId)
```

Comment:
I would pass StoreInterface instead of int (`$storeId`) to have more type safety